src/liballoc/boxed.rs

```diff
@@ -151,6 +151,43 @@ impl<T> Box<T> {
         Box(ptr.cast().into())
     }
 
+    /// Allocates memory on the heap *then* constructs `T` in-place.
+    ///
+    /// If the allocation fails, the initialization closure won't be called; because the allocation
+    /// can fail, the optimizer often couldn't construct `T` in-place even in theory because having
+    /// the allocation happen first is an observable change if the code that creates the value has
+    /// side-effects such as being able to panic.
+    ///
+    /// FIXME: This is intended to work via return-value-optimization, but due to compiler
+    /// limitations can't reliably do that yet.
+    #[inline(always)]
+    fn new_in_place(f: impl FnOnce() -> T) -> Box<T> {
```

Comment:
This is semantically more like `new_please_try_to_do_it_in_place_but_I_am_not_expecting_any_guarantee`?
(Such a guarantee is not something we can reliably or would make on a language (not LLVM) level.)

Reply:
Point is, if we had return-value-optimization, I think this is as close as you *could* get with an API that looked something like this given Rust's semantics. Even with out-references it's not clear to me you could get semantics that nicely guarantee no copies in an easy way, as you could still mess up and wind up with a copy prior to writing it to the out-reference.

Like I said above, it's not clear to me this is actually a good idea. But it does seem novel. :)

Reply:
Given that we do not guarantee copy elision we'll need to at minimum call this something other than `new_in_place` which suggests that there actually is a guarantee. For example: `new_hint_in_place`.

It might be a good idea to just use this internally for now... let's see how this measures up in terms of perf for `fn default`.

Reply:
> It might be a good idea to just use this internally for now

Yup, it is private right now, as I didn't want to give people false hope even in nightly. I'd suggest that we leave it that way until RVO is more reliable (specifically named-return-value-optimization if I understand the terminology right?).